service/storage_service.cc

```diff
@@ -1246,9 +1282,11 @@ class topology_coordinator {
             bool sleep = false;
             try {
                 auto guard = co_await start_operation();
-
                 co_await publish_oldest_cdc_generation(std::move(guard));
 
+                guard = co_await start_operation();
+                co_await clean_obsolete_cdc_generations(std::move(guard));
```

Comment:
Can we do both under one guard (one write)?

Reply:
We can but it will require some refactoring.

One way to do this is making `publish_oldest_cdc_generation` take guard by const reference and return mutation.

Another way is making `publish_oldest_cdc_generation` and `clean_obsolete_cdc_generations` take guard and vector of mutations by reference. Then, we could update the topology state directly in the publisher's loop.

Reply:
Ah, I didn't see your comment.

Still, I like two guards more TBH.

Reply:
Two guards are two barriers. Two chances to fail. The only advantage is two history rows, but we can be smart and build one entry that described what was done exactly.

Reply:
Ok. We could build an update string, like we build the update vector.

Reply:
Building the `reason` string may be a detail but I'll ask since I'm not sure how to approach it. I think we would like this building process to meet 2 requirements:
- the string should always look naturally (without something like ", " at the end),
- it shouldn't depend on the order in which `clean_obsolete_cdc_generations`,  `publish_oldest_cdc_generation` and possible future functions are called (unless they impose some order, but right now they don't).

For now, I implemented the solution that only meets the first requirement because meeting both of them requires more logic and I'm not sure if it's worth it.

Reply:
I think it's fine.